akka24-conductr-client-lib/src/main/scala/com/typesafe/conductr/clientlib/akka/BundlesConnector.scala

```diff
@@ -0,0 +1,265 @@
+    import JsonMarshalling._
+    Source
+      .single(request)
+      .via(connection)
+      .mapAsync(1)(Unmarshal(_).to[Seq[Bundle]])
+      .runForeach(connector ! GetBundlesSuccess(_))
+      .recover {
+        case NonFatal(e) => connector ! GetBundlesError(e)
+      }
+  }
+
+  /*
+   * When consuming bundle events we request the retrieval of bundle info only when a connection
+   * to the events stream is established. This ensures that we don't have a race condition whereby
+   * we miss events. Otherwise, if we retrieved the bundles before being connected to the event
+   * stream then we could miss out on events. We also cannot retrieve the bundle info only when
+   * having received events because then startup could fail if there are none i.e. we must always
+   * get the initial state of bundle info.
+   */
+  private def consumeBundlesEvents(connector: ActorRef, connection: Connection, request: HttpRequest)(implicit ec: ExecutionContext, mat: ActorMaterializer): Unit =
+    Source
+      .single(request)
+      .via(connection)
+      .via(
+        Streams.sseFlow(
+          Streams.onSuccess(_ => connector ! GetBundles),
+          outcome => connector ! ConsumeBundlesEventsOutcome(outcome)
+        )
+      )
+      .runForeach(connector ! _)
+
+}
+
+/**
+ * Connects to the `/bundles` endpoint of ConductR and publishes the bundles from ConductR when updates from
+ * the /bundles/events` SSE are available.
+ *
+ * If `stopAfter` is specified, [[BundlesConnector]] will terminate with [[TimeoutException]] if the bundles stream
+ * is not completed within the duration specified by `stopAfter`.
+ *
+ * Upon startup, [[BundlesConnector]] will obtain the bundle state from the `/bundles` endpoint, and publish these
+ * bundles before proceeding to connect to the `/bundle/events` endpoint of ConductR.
+ */
+class BundlesConnector(
+    conductrAddress: URL,
+    getBundlesEventsRequest: HttpRequest,
+    getBundlesRequest: HttpRequest,
+    stopAfter: Option[FiniteDuration],
+    expectingEventBurstTimeout: FiniteDuration,
+    reconnectInterval: FiniteDuration
+) extends FSM[BundlesConnector.State, Seq[Bundle]] with ActorPublisher[Seq[Bundle]] {
+
+  private implicit val mat = ActorMaterializer()
+
+  import BundlesConnector._
+  import context.dispatcher
+  import context.system
+
+  startWith(Initializing, Seq.empty)
+
+  stopAfter.foreach(context.system.scheduler.scheduleOnce(_, self, BundlesConnector.TimeoutException))
+
+  when(Initializing, reconnectInterval) {
+    case Event(GetBundles, _) =>
+      log.debug("Retrieving bundle info")
+      getBundles(self, createBundlesConnection(conductrAddress), getBundlesRequest)
+      stay()
+
+    case Event(GetBundlesSuccess(newBundles), _) =>
+      log.debug("Received bundles from ConductR: {}", newBundles)
+      if (totalDemand > 0) {
+        onNext(newBundles)
+        self ! ConsumeBundlesEvents
+        goto(ExpectingEvent).using(newBundles)
```

Comment:
Do we need to set the `newBundles` as state data? In `ExpectingEvent` we do not access this data. If we need this data for events that are part of `whenUnhandled` then we should think about moving these events from `whenUnhandled` to the corresponding states so that they are handled explicitly. Otherwise, the code is hard to read and to maintain.